### PR TITLE
flash attention output   bias grad

### PR DESCRIPTION
**Pull request by ltqin**

### CODE REVIEW

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp [hunk 1]
```diff
@@ -1359,13 +1359,13 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_V2
         static_assert(NPerXdl == 32);
         static_assert(D0BlockTransferSrcScalarPerVector * NThreadClusterLengths <= NPerBlock,
                       "D0BlockTransferSrcScalarPerVector * NThreadClusterLengths <= NPerBlock");
-        __host__ __device__ static constexpr auto GetD0BlockWriteDescriptor_M0_N0_M1_M2_N1_M3()
+        __host__ __device__ static constexpr auto GetD0BlockGlobalDescriptor_M0_N0_M1_M2_N1_M3()
         {
             // B1 matrix in LDS memory, dst of blockwise copy
```
Comment (danyao12) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp]: Modify this comment?
// B1 matrix in LDS memory, dst of blockwise copy

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp [hunk 2]
```diff
@@ -1303,13 +1303,13 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_V1
         static_assert(MPerXdl <= KPerBlock);
         static_assert(D0BlockTransferSrcScalarPerVector * NThreadClusterLengths <= NPerBlock,
                       "D0BlockTransferSrcScalarPerVector * NThreadClusterLengths <= NPerBlock");
-        __host__ __device__ static constexpr auto GetD0BlockWriteDescriptor_M0_N0_M1_M2_N1_M3()
+        __host__ __device__ static constexpr auto GetD0BlockGlobalDescriptor_M0_N0_M1_M2_N1_M3()
         {
             // B1 matrix in LDS memory, dst of blockwise copy
```
Comment (danyao12) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp]: Modify this comment?
// B1 matrix in LDS memory, dst of blockwise copy

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp [hunk 3]
```diff
@@ -1359,13 +1359,13 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_V2
         static_assert(NPerXdl == 32);
```
Comment (danyao12) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp:1361]: It looks like there will be bankconflict that requires further discussion.

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp [hunk 4]
```diff
@@ -1303,13 +1303,13 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_V1
         static_assert(MPerXdl <= KPerBlock);
```
Comment (danyao12) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp]: It looks like there will be bankconflict that requires further discussion.

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp (hunk 2 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp]: done

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp (hunk 1 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp]: done

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp (hunk 3 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp:1361]: I tried other non conflicting solutions, but the performance is not good

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp (hunk 4 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v1.hpp]: I tried other non conflicting solutions, but the performance is not good

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp (hunk 3 above)
Comment (danyao12) [reply include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp:1361]: Can you show us your codes, maybe using another branch.

**Review by qianfengz — COMMENTED** [review]
```C++ 
// YGrad in Gemm A position
    static auto MakeYGradGridDescriptor_O0_M_O1(const std::vector<index_t>& y_gs_ms_os_lengths,
                                                const std::vector<index_t>& y_gs_ms_os_strides)
    {
        return Transform::MakeAGridDescriptor_AK0_M_AK1(
            Transform::MakeAGridDescriptor_M_K(y_gs_ms_os_lengths, y_gs_ms_os_strides),
            Number<Y_O1>{});
    }
```

Is the above codes correct in file `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v1.hpp`  and `device_batched_mha_bwd_xdl_cshuffle_qloop_v1.hpp` ?

Also this codes need be removed in `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp` and `device_batched_mha_bwd_xdl_cshuffle_qloop_v2.hpp`

**Comment by danyao12** [comment]
> ```c++
> Transform::MakeAGridDescriptor_M_K(y_gs_ms_os_lengths, y_gs_ms_os_strides),
> ```

**Review by qianfengz — COMMENTED** [review]
```c++
     //
 597     // dS_i_j = P_i_j .* (dP_i_j - dY_i dot Y_i)
 598     //
 599 
 600     //
 601     // dQ = alpha * dS * K
 602     //
 603 
 604     // QGrad in Gemm C position
 605     static auto MakeQGradGridDescriptor_M_K(const std::vector<index_t>& q_gs_ms_ks_lengths,
 606                                             const std::vector<index_t>& q_gs_ms_ks_strides)
 607     {
 608         return Transform::MakeCGridDescriptor_M_N(q_gs_ms_ks_lengths, q_gs_ms_ks_strides);
 609     }
 610 
 611     //
 612     // dK = alpha * dS^T * Q
 613     //
 614 
 615     // KGrad in Gemm C position
 616     static auto MakeKGradGridDescriptor_N_K(const std::vector<index_t>& k_gs_ns_ks_lengths,
 617                                             const std::vector<index_t>& k_gs_ns_ks_strides)
 618     {
 619         return Transform::MakeCGridDescriptor_M_N(k_gs_ns_ks_lengths, k_gs_ns_ks_strides);
 620     }
```
Remove these codes since they are not used in files `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v1.hpp`, `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp`, `device_batched_mha_bwd_xdl_cshuffle_qloop_v1.hpp` and `device_batched_mha_bwd_xdl_cshuffle_qloop_v2.hpp`

**Comment by danyao12** [comment]
> ```c++
> // YGrad in Gemm A position
>     static auto MakeYGradGridDescriptor_O0_M_O1(const std::vector<index_t>& y_gs_ms_os_lengths,
>                                                 const std::vector<index_t>& y_gs_ms_os_strides)
>     {
>         return Transform::MakeAGridDescriptor_AK0_M_AK1(
>             Transform::MakeAGridDescriptor_M_K(y_gs_ms_os_lengths, y_gs_ms_os_strides),
>             Number<Y_O1>{});
>     }
> ```
> 
> Is the above codes correct in file `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v1.hpp` ?
> 
> Also this codes need be removed in `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp`

Correct codes.

**Comment by qianfengz** [comment]
> ```c++
> // YGrad in Gemm A position
>     static auto MakeYGradGridDescriptor_O0_M_O1(const std::vector<index_t>& y_gs_ms_os_lengths,
>                                                 const std::vector<index_t>& y_gs_ms_os_strides)
>     {
>         return Transform::MakeAGridDescriptor_AK0_M_AK1(
>             Transform::MakeAGridDescriptor_M_K(y_gs_ms_os_lengths, y_gs_ms_os_strides),
>             Number<Y_O1>{});
>     }
> ```
> 
> Is the above codes correct in file `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v1.hpp` and `device_batched_mha_bwd_xdl_cshuffle_qloop_v1.hpp` ?
> 
> Also this codes need be removed in `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp` and `device_batched_mha_bwd_xdl_cshuffle_qloop_v2.hpp`

Suggest to change `Transform::MakeAGridDescriptor_M_K` to `Transform::MakeCGridDescriptor_M_N`

**Comment by qianfengz** [comment]
> > ```c++
> > // YGrad in Gemm A position
> >     static auto MakeYGradGridDescriptor_O0_M_O1(const std::vector<index_t>& y_gs_ms_os_lengths,
> >                                                 const std::vector<index_t>& y_gs_ms_os_strides)
> >     {
> >         return Transform::MakeAGridDescriptor_AK0_M_AK1(
> >             Transform::MakeAGridDescriptor_M_K(y_gs_ms_os_lengths, y_gs_ms_os_strides),
> >             Number<Y_O1>{});
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is the above codes correct in file `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v1.hpp` ?
> > Also this codes need be removed in `device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp`
> 
> Correct codes.

Change `Transform::MakeAGridDescriptor_M_K` to `Transform::MakeCGridDescriptor_M_N` for readability.    We can also rename all `C` to `C1`  since we already have `C0`

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp [hunk 5]
```diff
@@ -1273,34 +1276,77 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_Light_V1
...
-            4,                                               // SrcVectorDim
-            5,                                               // DstVectorDim
-            D0BlockTransferSrcScalarPerVector,               // SrcScalarPerVector
-            4,                                               // DstScalarPerVector
+                     1>,                                      // ThreadClusterLengths
+            Sequence<0, 1, 2, 3, 5, 4>,                       // ThreadClusterArrangeOrder
+            typename TypeTransform<D0DataType>::Type,         // SrcData
+            typename TypeTransform<D0DataType>::Type,         // DstData
+            D0GridDescriptor_M0_N0_M1_M2_N1_M3,               // SrcDesc
+            decltype(d0_block_global_desc_m0_n0_m1_m2_n1_m3), // DstDesc
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp]: Rename this to `d0_block_dst_desc_m0_n0_m1_m2_n1_m3`

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp [hunk 6]
```diff
@@ -1273,34 +1276,77 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_Light_V1
...
             true,
             true, // DstResetCoord
             1>;
 
-        using D0ThreadWiseCopy =
+        using D0ThreadwiseCopyLdsToVgpr =
             ThreadwiseTensorSliceTransfer_v4<typename TypeTransform<D0DataType>::Type,    // SrcData
                                              typename TypeTransform<D0DataType>::Type,    // DstData
-                                             decltype(d0_block_read_desc_n0_n1_m0_m1_m2), // SrcDesc
+                                             decltype(d0_block_vgpr_desc_n0_n1_m0_m1_m2), // SrcDesc
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp]: Rename this to `d0_block_src_desc_n0_n1_m0_m1_m2`

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp [hunk 7]
```diff
@@ -1273,34 +1276,77 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_Light_V1
...
+            tensor_operation::element_wise::Scale, // CElementwiseOperation
+            Sequence<1, 1, 4, 1, 4>,               // SliceLengths
+            Sequence<0, 1, 2, 3, 4>,               // AccessOrder
+            4,                                     // VectorDim
+            4,                                     // ScalarPerVector
+            InMemoryDataOperationEnum::Set,        // GlobalMemoryDataOperation
+            1,                                     // DstScalarStrideInVector
+            true>;
+
+        using D0BlockwiseCopyLdsToGlobal = ThreadGroupTensorSliceTransfer_v4r1<
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp]: Rename `D0BlockwiseCopyLdsToGloba`  to `D0GradBlockwiseCopyLdsToGlobal`

File: include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp [hunk 8]
```diff
@@ -1273,34 +1276,77 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_Light_V1
...
+            Sequence<1,
+                     1,
+                     1,
+                     BlockSize / NThreadClusterLengths,
+                     NThreadClusterLengths,
+                     1>,                                      // ThreadClusterLengths
+            Sequence<0, 1, 2, 3, 5, 4>,                       // ThreadClusterArrangeOrder
+            typename TypeTransform<D0DataType>::Type,         // SrcData
+            typename TypeTransform<D0DataType>::Type,         // DstData
+            decltype(d0_block_global_desc_m0_n0_m1_m2_n1_m3), // SrcDesc
```
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_light_v1.hpp]: Add `auto & d0grad_block_src_desc_m0_n0_m1_m2_n1_m3` = `d0_block_dst_desc_m0_n0_m1_m2_n1_m3` if they are completely same in the LDS layout

**Review by qianfengz — COMMENTED** [review]
Using type aliasing and object reference reuse codes involving d0 and d0_grad

**Review by qianfengz — APPROVED** [review]

**Pull request merged**